README.md
# There are SLIMES in this DUNGEON
This is the code for my 2020 7DRL attempt: There are SLIMES in this DUNGEON  
Result: SUCCESS!

## Controls
W, A, S, D Keys to Move  
Number Keys to use special abilities (when unlocked)   
Spacebar to pass your turn  
Collect the Crystals to Escape!

## Credits
Design, Programming, Art: nuzcraft  
Sprites: Quale  (https://imgur.com/a/uHx4k)  
Background Music: Exit Only by Lately Kind of Yeah (freemusicarchive.org)  
Sounds Effects made with https://www.bfxr.net  

https://github.com/nuzcraft/slimes_in_this_dungeon

## Special Thanks
[humbit](https://twitter.com/humbit) for their amazing JS Broughlike tutorial (https://nluqo.github.io/broughlike-tutorial/)  
[kyzrati](https://twitter.com/kyzrati) for getting me into roguelikes and being an awesome person

## Postmortem
I'll try to keep this short. 😉  
I had an amazing experience developing this game over the last 7 days. I don't have a lot of spare time, so I worked hard to get as much planned out as possible ahead of time and it turned out so well! In the end, There are SLIMES in this DUNGEON was developed in ~16 hours, something only possible because my project plan minimized downtime. Check out the repo on github for the full Project Plan, Changelog, and code. 10/10, will participate again!

As with all projects, there are a bunch of improvements I'd like to make. Below are a few stretch goals that didn't make it in due to the time constraints:  
* Story - this could have been anything!
* Advanced AI - the current slime AI is nothing to write home about. They kill themselves on EVERYTHING.
* Improved Abilities - with a few small tweaks (aka some TLC) all the abilities could be way more interesting and balanced.
* More Levels - the game is so short!! I had 3 additional levels planned, but even then it would be too short.
* Animations - just for fun 

While developing, I started thinking about the progression, and feel like that could be improved a lot too. With more levels, it would have been cool to let the play choose between 2 or 3 possible next levels. That way, they could have some say over the next type of enemy or the next gained ability. Hindsight is 20/20!!

Regardless, I'm very happy with the result!


CHANGELOG.md
## Changelog
### 2020.03.06 Day 7, the final day!
* 11:30 AM, lets' work on a title screen!
* So, I made some really bad art that I really like for for a title screen.
* 1 hr
* 8:OO PM...Ha! one of the first nights that I get to start early and I'm pretty much done! Gonna work on the game over and win screens, then a short writeup for the readme.
* Alright! I made super basic 'you died' and 'thanks for playing' screens to show at the end.
* 1 hr
* Total Dev Time = 2 hrs
### 2020.03.05 Day 6!
* 6:50 AM, I've got 10 mins till work starts. I'm gonna try to switch the wind gusts to act more like walls, and allow the player to place them around them instead of firing them at enemies.
* Okay, I'm much happier with this setup, it feels less like a cop out and a copy of throwing poison bombs.
* 20 mins
* 11:30 AM, over my lunch, time to get cracking
* Hurrah!! We are functionally done!! I've hit everything in my main project plan! Now, it's time to plan some stretch goals :D. 
* Hello v1.0!!
* 1 hr
* 8:45, alright, time to polish some stuff up!
* I got some sounds and music mixed up. Background effects from https://www.bfxr.net/ and a background song from https://files.freemusicarchive.org/storage-freemusicarchive-org/music/Decoder_Magazine/Lately_Kind_of_Yeah/Summit/Lately_Kind_of_Yeah_-_04_-_Exit_Only.mp3
* Exit Only by Lately Kind of Yeah
* I added some text to the bottom of the screen giving some really basic instructions. I also had to add a secret command...space bar to cause the game to tick. This makes it WAY easier to win, and I don't like it. It's basically just there in case you get stuck in a corner.
* Tomorrow, I'm going to update the title screen and add a game over screen, and we'll probably call it done!
* 2.25 hrs
* Total Dev Time: 3.5 hrs
### 2020.03.04 Day 5!
* 1:37 PM, HOLY CANOLI DAY 5 ALREADY. Also, I've not been able to do ANY work on this today (besides write this note). My fingers are crossed that I'll get time tonight... I'm definitely glad that I structured my project plan so that I could piecemeal this out...I'll be lucky to finish the cyan and yellow levels by the end at this rate (which was my minimum viable product!!)
* In my mind, cyan and yellow should be quicker, but I could run into issues.
* 9 PM, once again, lots of chores. But, I've got an IPA and am ready to rock!
* Fuck, that felt like a struggle. For now, the Magenta level is complete!!
* Okay, I doubled down, spent a little extra time, and got the Cyan level finished! I'm not super stoked about how the wind gusts work, but I might have time to fix them later (it might be easy if they just push the monster or player to their previous position, AKA a wall with a tick). Tomorrow is Yellow day + the start of polish.
* NOTE: there is a bug where things like poison clouds and bombs can delete crystals and exits, which is bad and should be fixed.
* Total Dev Time - 2 hrs
### 2020.03.03 Day 4!
* 6:30, got to work early again!
* made some progress getting poison clouds working...I may make them their own tile similar to crystals, not sure tho
* 45 mins
* 9 PM. I had a lot of chores to do before I could get started. Honestly, this 7DRL couldn't have fallen on a worse week! We are packing up the house to move! Super exciting! but also, lots of fucking work.
* Things are starting to slow... I got most of the logic set up for poison bombs and clouds, but still have to set it up so that enemies and players can use them. It's kind of amazing, but just adding the poison clouds has made the magenta level so much more dynamic. Tomorrow, I'll finish up the magenta level and start on Cyan.
* 1.5 hrs
* Total Dev Time - 2.25 hrs
### 2020.03.02 Day 3!
* 6:40, got to work a bit early, will hopefully get 20-30 mins of work in, then an hour or so at lunch.
* 20 mins, made a little progress, but nothing to mention.
* Okay, so, I had to leave work early and work from home because I barfed 🤮 I'm not sick! My kids had the flu last week and I'm on Tamiflu to keep from being infected and it apparently doesn't sit well with an empty stomach...it was awful. So, I worked through lunch instead, I'll have to try to make up this time tonight. The goal: complete BSP level generation!
* 8:40 PM, getting started again!
* Wow, really good progress tonight. Level gen is complete! There will likely be some tweaks, I'm probably going to try to expand the map size to accomodate everything, but it looks great for now!
* rooms are created, and enemies are spawned in rooms. The player is placed in one room with no enemies, and the crystal is placed in a different room
* 1 hr 20 mins
* Total Dev Time: 1 hr 40 min
### 2020.03.01 Day 2!
* 5:00 AM, up early! I was hoping to get some dev time in before the family wakes up, and it worked out!
* I finished a couple sections of the project plan:
* added enemies
* gave them simple ai
* added combat (one hit kills)
* added level exits
* TECHNICALLY it's now a game! It has a start, gameplay with a goal, and an end. There's lots more to do though! The next step is to add the collectible crystals and the gameplay mechanic where they will unlock the stairs.
* 1.25 hrs
* 8:30 PM (ish) I'm back!! I can't remember where we're starting tho 😅
* aight! great progress tonight!! I got it set up so the player can collect crystals and use them to advance to the next level. There are currently 3 crystals to collect then a final floor to escape the dungeon. Since enemy ai is pretty simple and the player can only bump attack, its actually pretty hard to win!
* I'll be uploading a Day 2 build to itch.io today as well.
* 2 hours
* Total Dev Time: 3.25 hours
### 2020.02.29 Day 1 begins!
* 5:30 AM, 7DRL has begun!!
* 7 WHOLE MINUTES OF WORK before I had to stop to help deal with a kid 😑
* I made good progress for only ~45 mins of time spent. I've got a working title screen and the start of something good! I grabbed the spritesheet from Row Jam Brough for now, but that will be replaced with a new one when I get there. So exciting!!

js/map.js

function generateLevel(color){
    // start by generating tiles on the map
    generateTiles();

    // generate monsters here
    generateMonsters();

    // generate traps here (aka non-monsters)
    generateTraps();

    // generate other stuff here
    if (level < numLevels){
        generateCrystal(color);
    } else {
        randomPassableTile().replace(Exit);
    }
    
}

function generateTiles(){
    // this function will be used to generate the tiles of the map
    let passableTiles = 0;
    let minXInBounds = numTiles_x;
    let minYInBounds = numTiles_y;
    let width = 0;
    let height = 0;

    // start by building an empty room
    // get the dimensions of the room as well
    tiles = [];
    for (let i = 0; i < numTiles_x; i++){
        tiles[i] = [];
        for (let j = 0; j < numTiles_y; j++){
            if (!inBounds(i, j)){
                tiles[i][j] = new Wall(i, j);
            } else {
                minXInBounds = ((i <= minXInBounds) ? i : minXInBounds)
                minYInBounds = ((j <= minYInBounds) ? j : minYInBounds)
                width = i - minXInBounds + 1;
                height = j - minYInBounds + 1;
                tiles[i][j] = new Floor(i, j);
                passableTiles++;
            }
        }
    }

    // create a container the size of the room
    let main_container = new Container(minXInBounds, minYInBounds, width, height);
    // console.log(main_container);
    let container_tree = split_container(main_container, 3);
    // console.log(container_tree);

    // add all the leaves to an array
    rooms = shuffle(container_tree.getLeafs())
    // console.log(rooms);

    // now, let's find all the space included in the tree and mark anything
    // not in those spaces as walls
    let points = [];
    points = container_tree.getIncludedPoints(points);

    // now let's connect the rooms and add those hallways to the 'points'
    for(let i = 0; i < rooms.length - 1; i++) {
        let center1 = rooms[i].center
        let center2 = rooms[i+1].center
        let horiz = randomRange(0, 1);

        // if horizontal, start with a horizontal tunnel
        if (horiz == 1){
            let x1 = Math.min(center1.x, center2.x);
            let x2 = Math.max(center1.x, center2.x);
            let y = center1.y;
            for (let j = x1; j <= x2; j++){
                points.push([j, y]);
            }
            let y1 = Math.min(center1.y, center2.y);
            let y2 = Math.max(center1.y, center2.y);
            let x = center2.x;
            for (let k = y1; k <= y2; k++){
                points.push([x, k]);
            }
        } else if (horiz == 0){
            let y1 = Math.min(center1.y, center2.y);
            let y2 = Math.max(center1.y, center2.y);
            let x = center1.x;
            for (let k = y1; k <= y2; k++){
                points.push([x, k]);
            }
            let x1 = Math.min(center1.x, center2.x);
            let x2 = Math.max(center1.x, center2.x);
            let y = center2.y;
            for (let j = x1; j <= x2; j++){
                points.push([j, y]);
            }
        }
    }
   
    // now we actually create walls and floors
    for(let i=0;i<numTiles_x;i++){
        tiles[i]=[];
        for(let j=0;j<numTiles_y;j++){
            let inPoints = false;
            for (let k = 0; k < points.length; k++){
                if (points[k][0] === i && points[k][1] ===j){
                    inPoints = true;
                }
            }
            if(!inBounds(i, j) || !inPoints){ // removed the 30% for a wall
                tiles[i][j] = new Wall(i, j);
            }else{
                tiles[i][j] = new Floor(i, j);
                passableTiles++;
            }
        }
    }

    return passableTiles;
}

function inBounds(x, y){
    // thiw will return true if inside the border
    return x > 0 && y > 0 && x < numTiles_x - 1 && y < numTiles_y - 1;
}

function getTile(x, y, color){
    if(inBounds(x, y)){
        return tiles[x][y];
    } else {
        var wall = new Wall(x, y);
        return wall;
    }
}

function randomPassableTile(){
    let tile;
    tryTo('get random passable tile', function(){
        let x = randomRange(0, numTiles_x - 1);
        let y = randomRange(0, numTiles_y - 1);
        tile = getTile(x, y);
        return tile.passable && !tile.monster && !tile.isExit && !tile.isCrystal;
    })
    return tile;
}

function randomPassableTileInRoom(room){
    let tile;
    tryTo('get random passable tile', function(){
        let x = randomRange(room.x, room.x + room.w - 1);
        let y = randomRange(room.y, room.y + room.h - 1);
        tile = getTile(x, y);
        return tile.passable && !tile.monster && !tile.isExit && !tile.isCrystal;
    })
    return tile;
}

function generateMonsters(){
    monsters = [];
    // start with i = 1 bc room 1 is where the player will start
    for(let i = 1; i < rooms.length; i++){
        let numMonsters = randomRange(0, 2);
        for(let j = 0; j < numMonsters; j++){
            spawnMonsterInRoom(rooms[i], level_color);
        }
    }
}

function spawnMonster(){
    let possibleMonsters = collected_crystals.slice();
    possibleMonsters.push('colorless');
    let chosenColor = shuffle(possibleMonsters)[0];
    if (chosenColor == 'magenta'){
        let monster = new MagentaSlime(randomPassableTile());
        monsters.push(monster); 
    } else if (chosenColor == 'cyan'){
        let monster = new CyanSlime(randomPassableTile());
        monsters.push(monster);
    } else if (chosenColor == 'yellow'){
        let monster = new YellowSlime(randomPassableTile());
        monsters.push(monster);
    } else {
        let monster = new Slime(randomPassableTile());
        monsters.push(monster);
    }
}

function spawnMonsterInRoom(room, color){
    let chosenColor = color;
    if  (chosenColor === undefined){
        let possibleMonsters = collected_crystals.slice();
        possibleMonsters.push('colorless');
        chosenColor = shuffle(possibleMonsters)[0];
    }
    if (chosenColor == 'magenta'){
        let monster = new MagentaSlime(randomPassableTileInRoom(room));
        monsters.push(monster); 
    } else if (chosenColor == 'cyan'){
        let monster = new CyanSlime(randomPassableTileInRoom(room));
        monsters.push(monster); 
    } else if (chosenColor == 'yellow'){
        let monster = new YellowSlime(randomPassableTileInRoom(room));
        monsters.push(monster); 
    } else {
        let monster = new Slime(randomPassableTileInRoom(room));
        monsters.push(monster);
    }
}

function generateCrystal(color){
    // add a switch so that we pass in the right color
    // for the right crystal

    // choose the room to spawn the crystal in, not in room[0]
    let i = randomRange(1, rooms.length - 1);
    if (color == 'magenta') {
        randomPassableTileInRoom(rooms[i]).replace(MagentaCrystal);
    } else if (color == 'cyan'){
        randomPassableTileInRoom(rooms[i]).replace(CyanCrystal);
    } else if (color == 'yellow'){
        randomPassableTileInRoom(rooms[i]).replace(YellowCrystal);
    }
}

function generateTraps(){
    for (let i = 1; i < rooms.length; i++){
        if (level_color == 'magenta' || level_color === undefined){
            let numPoisonClouds = randomRange(0, 2);
            for (let j = 0; j < numPoisonClouds; j++){
                let tile = randomPassableTileInRoom(rooms[i]);
                tiles[tile.x][tile.y] = new PoisonCloud(tile.x, tile.y, randomRange(5, 10));
            }
        } 
        if (level_color == 'cyan' || level_color === undefined){
        let numSpikes = randomRange(0, 2);
            for (let j = 0; j < numSpikes; j++){
                let tile = randomPassableTileInRoom(rooms[i]);
                tiles[tile.x][tile.y] = new Spikes(tile.x, tile.y, randomRange(4, 6));
            }
        }
    }
}


js/tile.js
class Tile{
    constructor(x, y, sprite, passable){
        this.x = x;
        this.y = y;
        this.sprite = sprite;
        this.passable = passable;
    }

    draw(){
        drawSprite(this.sprite, this.x, this.y);
    }

    getNeighbor(dx, dy){
        return getTile(this.x + dx, this.y + dy);
    }

    getAdjacentNeighbors(){
        return shuffle([
            this.getNeighbor(0, -1),
            this.getNeighbor(0, 1),
            this.getNeighbor(-1, 0),
            this.getNeighbor(1, 0)
        ]);
    }

    getAdjacentPassableNeighbor(){
        return this.getAdjacentNeighbors().filter(t => t.passable).filter(v => !v.isExit).filter(w => !w.isCrystal);
    }

    getAdjacentPassableNeighborWithoutMonsters(){
        return this.getAdjacentNeighbors().filter(t => t.passable).filter(u => !u.monster).filter(v => !v.isExit).filter(w => !w.isCrystal);
    }

    // manhattan distance
    dist(other){
        return Math.abs(this.x - other.x) + Math.abs(this.y - other.y);
    }

    replace(newTileType){
        tiles[this.x][this.y] = new newTileType(this.x, this.y);
        return tiles[this.x][this.y];
    }

    setEffect(effectSprite){
        this.effect = effectSprite;
        this.effectCounter = 30;
    }

    drawEffect(){
        if (this.effectCounter){
            this.effectCounter--;
            ctx.globalAlpha = this.effectCounter / 30;
            drawSprite(this.effect, this.x, this.y);
            ctx.globalAlpha = 1;
        }
    }
}

class Floor extends Tile {
    constructor(x, y){
        super(x, y, spr_floor, true);
    }

    stepOn(monster){
        // does nothing for now
    }
}

class Wall extends Tile {
    constructor(x, y){
        super(x, y, spr_wall, false);
        // set the sprite based on the color of the level
        if (level_color == 'magenta'){
            this.sprite = spr_magenta_wall;
        } else if (level_color == 'cyan'){
            this.sprite = spr_cyan_wall;
        } else if (level_color == 'yellow'){
            this.sprite = spr_yellow_wall;
        }
    }
}

class Exit extends Tile{
    constructor(x, y){
        super(x, y, spr_exit_level, true);
        if (level == numLevels){
            this.sprite = spr_exit_game;
        }
        this.isExit = true;
    }

    stepOn(monster){
        if(monster.isPlayer){
            if(level == numLevels){
                showWinScreen();
                // showTitle();
            } else {
                level ++;
                startLevel();
                playSound("new_level");
            }
        }
    }
}

class MagentaExit extends Tile{
    constructor(x, y){
        super(x, y, spr_magenta_exit, true);
        this.isExit = true;
    }

    stepOn(monster){
        if(monster.isPlayer){
            if(level == numLevels){
                showWinScreen();
                // showTitle();
            } else {
                level ++;
                startLevel();
                playSound("new_level");
            }
        }
    }
}

class CyanExit extends Tile{
    constructor(x, y){
        super(x, y, spr_cyan_exit, true);
        this.isExit = true;
    }

    stepOn(monster){
        if(monster.isPlayer){
            if(level == numLevels){
                showWinScreen();
                // showTitle();
            } else {
                level ++;
                startLevel();
                playSound("new_level");
            }
        }
    }
}

class YellowExit extends Tile{
    constructor(x, y){
        super(x, y, spr_yellow_exit, true);
        this.isExit = true;
    }

    stepOn(monster){
        if(monster.isPlayer){
            if(level == numLevels){
                showWinScreen();
                // showTitle();
            } else {
                level ++;
                startLevel();
                playSound("new_level");
            }
        }
    }
}

class MagentaCrystal extends Tile {
    constructor(x, y){
        super(x, y, spr_magenta_crystal, true);
        this.isCrystal = true;
    }

    stepOn(monster){
        // spawn the exit
        if (monster.isPlayer){
            randomPassableTile().replace(MagentaExit);
            // remove the crystal from the floor
            this.replace(Floor);
            this.monster = monster;
            // add the crystal to the player
            monster.hasMagenta = true;
            collected_crystals.push("magenta");
            playSound("collect_gem");
        }
    }
}

class CyanCrystal extends Tile {
    constructor(x, y){
        super(x, y, spr_cyan_crystal, true);
        this.isCrystal = true;
    }

    stepOn(monster){
        // spawn the exit
        if (monster.isPlayer){
            randomPassableTile().replace(CyanExit);
            // remove the crystal from the floor
            this.replace(Floor);
            this.monster = monster;
            // add the crystal to the player
            monster.hasCyan = true;
            collected_crystals.push("cyan");
            playSound("collect_gem");
        }
    }
}

class YellowCrystal extends Tile {
    constructor(x, y){
        super(x, y, spr_yellow_crystal, true);
        this.isCrystal = true;
    }

    stepOn(monster){
        // spawn the exit
        if (monster.isPlayer){
            randomPassableTile().replace(YellowExit);
            // remove the crystal from the floor
            this.replace(Floor);
            this.monster = monster;
            // add the crystal to the player
            monster.hasYellow = true;
            collected_crystals.push("yellow");
            playSound("collect_gem");
        }
    }
}

class PoisonCloud extends Floor {
    constructor(x, y, timer=null){
        super(x, y);
        this.sprite = spr_poison_cloud;
        this.timer = timer;
    }

    stepOn(monster){
        monster.hit();
    }

    timesUp(){
        // destroy this poison cloud
        this.replace(Floor);
        // if not from a bomb, spawn a new cloud
        if (!this.fromBomb){
            let tile = randomPassableTile();
            tiles[tile.x][tile.y] = new PoisonCloud(tile.x, tile.y, randomRange(5, 10));
        }
    }
}

class WindGust extends Wall{
    constructor(x, y, timer=null){
        super(x, y);
        this.sprite = spr_wind_gust;
        this.timer = timer;
        playSound("cast_tornado");
    }

    timesUp(){
        // destroy this poison cloud
        this.replace(Floor);
    }
}

class Spikes extends Floor {
    constructor(x, y, timer=null){
        super(x, y);
        this.sprite = spr_spikes;
        this.timer = timer;
        this.retracted = false;
        this.resetTimer = timer;
    }

    stepOn(monster){
        if (!this.retracted){
            monster.hit();
        }
    }

    timesUp(){
        // if we use a time, use the timer to retract and extend
        // the spikes
        if (this.retracted){
            this.retracted = false;
            this.sprite = spr_spikes;
            this.timer = this.resetTimer;
        } else {
            this.retracted = true;
            this.sprite = spr_floor;
            this.timer = this.resetTimer;
        }
    }
}


js/tree.js
class Tree{
    constructor(leaf){
        this.leaf = leaf;
        this.lchild = undefined;
        this.rchild = undefined;
    }

    getLeafs(){
        if (this.lchild === undefined && this.rchild === undefined){
            return [this.leaf]
        } else {
            return [].concat(
                this.lchild.getLeafs(),
                this.rchild.getLeafs()
                )
        }
    }

    getIncludedPoints(points){
        if (this.lchild === undefined && this.rchild === undefined){
            points.concat(this.leaf.getIncludedPoints(points));
        }
        if (this.lchild !== undefined){
            points.concat(this.lchild.getIncludedPoints(points));
        }
        if (this.rchild !== undefined){
            points.concat(this.rchild.getIncludedPoints(points));
        }
        // if there is an lchild and rchild, create and add points to
        // connect the two centers
        // if (this.lchild !== undefined && this.rchild !== undefined){
        //     points.concat(this.lchild.leaf.pointsBetweenCenters(points, this.rchild.leaf));
        // }
        return points
    }
}

js/game.js

// colors
const color_white = "#ffffff";
const color_magenta = "#FF00FF";
const color_cyan = "#00FFFF";
const color_yellow = "#FFFF00";
const color_black = "#000000";
const color_dark_grey = "#555555";

// sprite indexes
const spr_player = 0;
const spr_wall = 1;
const spr_floor = 2;
const spr_slime = 3;
const spr_player_dead = 4;
const spr_exit_level = 5;
const spr_exit_game = 6;
const spr_magenta_crystal = 7;
const spr_cyan_crystal = 8;
const spr_yellow_crystal = 9;
const spr_magenta_exit = 10;
const spr_cyan_exit = 11;
const spr_yellow_exit = 12;
const spr_empty_crystal = 13;
const spr_magenta_wall = 14;
const spr_cyan_wall = 15;
const spr_yellow_wall = 16;
const spr_poison_cloud = 17;
const spr_empty = 18;
const spr_poison_bomb = 19;
const spr_magenta_slime = 20;
const spr_cyan_slime = 21;
const spr_yellow_slime = 22;
const spr_wind_gust = 23;
const spr_spikes = 24;
const spr_lightning = 25;
const spr_yellow_slime_charged = 26;

var level_colors = ['magenta', 'cyan', 'yellow'];
var level_color = 'magenta';
var collected_crystals = [];
var rooms = [];
var monsters = [];

function setupCanvas() {
    // console.log("setUpCanvas started");
    canvas = document.querySelector("canvas");
    ctx = canvas.getContext("2d");

    canvas.width = tileSize * numTiles_x;
    canvas.height = tileSize * (numTiles_y + uiHeight);
    canvas.style.width = canvas.width + 'px';
    canvas.style.height = canvas.height + 'px';
    ctx.imageSmoothingEnabled = false;
}

function drawText(text, size, centered, textX, textY, color){
    // console.log("drawTextStarted");
    ctx.fillStyle = color;
    ctx.font = size + "px serif";
    // ctx.font = size + "px monospace";
    if (centered) { // generally, print in center of whole screen
        textX = ((canvas.width - ctx.measureText(text).width) / 2) + textX;
    } else { // if not centered, put it in the ui section
        // textY = canvas.height - uiHeight * tileSize + 25 + textY;
    }
    ctx.fillText(text, textX, textY);
}

function showTitle(){
    // console.log("show title started");
    ctx.drawImage(title, 0, 0);
    ctx.fillStyle = 'rgba(0, 0, 0, .75)';
    ctx.fillRect(canvas.width / 3.5, canvas.height / 7, canvas.width / 2.33, canvas.height / 2.5);
    gameState = "title";
    drawText("There are", 40, true, 0, canvas.height / 2 - 170, color_white);
    drawText("SLIMES", 70, true, 0, canvas.height / 2 - 110, color_cyan);
    // drawText("in this DUNGEON", 40, true, 0, canvas.height / 2 - 70, color_white);
    drawText("in this                     ", 40, true, 0, canvas.height / 2 - 70, color_white);
    drawText("           DUNGEON", 40, true, 0, canvas.height / 2 - 70, color_magenta);
    drawText("Press any key to start", 20, true, 0, canvas.height / 2 - 35, color_yellow);
}

function showGameOver(){
    ctx.fillStyle = 'rgba(0, 0, 0, .75)';
    ctx.fillRect(0, 0, canvas.width, canvas.height);
    ctx.drawImage(gameover, 0, 0);
    gameState = 'gameover';
}

function showWinScreen(){
    ctx.fillStyle = 'rgba(0, 0, 0, .75)';
    ctx.fillRect(0, 0, canvas.width, canvas.height);
    ctx.drawImage(win, 0, 0);
    gameState = 'win';
}

function startGame(){
    console.log("Starting game.");
    level = 1;
    level_colors = ['magenta', 'cyan', 'yellow']
    collected_crystals = [];
    startLevel();
    gameState = "running";

    if (background_audio == 0){
        let background = new Audio('sounds/Lately_Kind_of_Yeah_-_04_-_Exit_Only.mp3');
        background.currentTime = 0;
        background.play();
        background.loop = true;
        background.volume = .5;
        background_audio = 1;
    }
}

function startLevel(){
    console.log("Starting level.");
    spawnRate = 15;
    spawnCounter = spawnRate;

    // get a color for this level
    level_color = shuffle(level_colors)[0];
    // console.log(level_color);
    // remove this color from the level array
    level_colors.splice(level_colors.indexOf(level_color), 1);
    // console.log(level_colors);

    generateLevel(level_color);

    // spawn the player in the first room
    player = new Player(randomPassableTileInRoom(rooms[0]));
    // player.hasYellow = true; // for debugging
    console.log("Player placed in level.");
    for (var i = 0; i < collected_crystals.length; i++){
        if (collected_crystals[i] == 'magenta'){
            player.hasMagenta = true;
        } else if (collected_crystals[i] == 'cyan'){
            player.hasCyan = true;
        } else if (collected_crystals[i] == 'yellow'){
            player.hasYellow = true;
        }
    }

    // this was moved into the StepOn code for crystals
    // randomPassableTile().replace(Exit);
}

function draw() {
    if (gameState == "running" || gameState == "dead"){
        ctx.clearRect(0, 0, canvas.width, canvas.height);

        screenshake();

        for(let i = 0; i < numTiles_x; i++){
            for (let j = 0; j < numTiles_y; j++){
                getTile(i, j, level_color).draw();
            }
        }

        for(let i = 0; i < monsters.length; i++){
            monsters[i].draw();
        }

        player.draw();

        // draw effects
        for(let i = 0; i < numTiles_x; i++){
            for (let j = 0; j < numTiles_y; j++){
                getTile(i, j, level_color).drawEffect();
            }
        }

        // draw ui
        if (player.hasMagenta){
            drawSprite(spr_magenta_crystal, 0, 9);
            let textColor = color_dark_grey;
            if (player.poisonBombCooldown == 0){
                textColor = color_magenta;
            }
            drawText("(1)", 28, false, 56, 9*64+36, textColor);
            drawText("Throw Poison Bomb",20, false, 90, 9*64+28, textColor);
            drawText("at nearest enemy", 20, false, 94, 9*64+44, textColor);
        } else {
            drawSprite(spr_empty_crystal, 0, 9);
        }

        if (player.hasCyan){
            drawSprite(spr_cyan_crystal, 0, 10);
            let textColor = color_dark_grey;
            if (player.windGustCooldown == 0){
                textColor = color_cyan;
            }
            drawText("(2)", 28, false, 56, 10*64+36, textColor);
            drawText("Summon Tornado",20, false, 90, 10*64+28, textColor);
            drawText("in front of you", 20, false, 98, 10*64+44, textColor);
            drawText("(based on last movement)", 14, false, 86, 10*64+57, textColor);
        } else {
            drawSprite(spr_empty_crystal, 0, 10);
        }

        if (player.hasYellow){
            drawSprite(spr_yellow_crystal, 4, 9);
            let textColor = color_dark_grey;
            if (player.lightningCooldown == 0){
                textColor = color_yellow;
            }
            drawText("(3)", 28, false, 5*64, 9*64+36, textColor);
            drawText("Summon LIGHTNING",20, false, 5*64+44, 9*64+28, textColor);
            drawText("horizontally and vertically", 20, false, 5*64+36, 9*64+44, textColor);
            // drawText("(based on last movement)", 14, false, 86, 10*64+57, textColor);
        } else {
            drawSprite(spr_empty_crystal, 4, 9);
        }

        // draw the instructions there at the bottom
        drawText("Use the W,A,S,D keys to move around.", 25, false, 4*64, 10*64+16, color_white);
        drawText("Use the number keys to use abilities.", 25, false, 4*64, 10*64+48, color_white);
        drawText("Collect the CRYSTALS", 18, false, 10*64+16, 9*64+28, color_white);
        drawText("and", 18, false, 11*64-16+8, 9*64+48, color_white);
        drawText("ESCAPE", 18, false, 11*64+16+8, 9*64+48, color_magenta);
        drawText("the", 18, false, 11*64-24+8, 10*64+4, color_white);
        drawText("DUNGEON", 18, false, 11*64+8, 10*64+4, color_cyan);
        drawText("of", 18, false, 11*64-8+8, 10*64+16+8, color_white);
        drawText("SLIMES", 18, false, 11*64+12+8, 10*64+16+8, color_yellow);
        // drawText("of !")
    }
}

function drawSprite(sprite, x, y) {
    ctx.drawImage(
        spritesheet,
        (sprite % 8) * 16,
        Math.floor(sprite / 8) * 16,
        16,
        16,
        x * tileSize + shakeX,
        y * tileSize + shakeY,
        tileSize,
        tileSize
    )
}

function tick(){
    // let monsters take their turn / remove dead monsters
    for(let k=monsters.length-1;k>=0;k--){
		if(!monsters[k].dead){
            monsters[k].update();
            // monsters[k].throwPoisonBomb();
            if (monsters[k].timer){
                monsters[k].timer--;
                if (monsters[k].timer == 0){
                    monsters[k].timesUp();
                }
            }
		}else{
			monsters.splice(k, 1);
		}
    }

    // player.update();
    if (player.dead){
        gameState = "dead";
    }

    // spawn new monsters in the level
    spawnCounter--;
    if (spawnCounter <= 0){
        spawnMonster();
        spawnCounter = spawnRate;
        spawnRate --;
    }

    // check for tiles that run on a timer
    for(let i = 0; i < numTiles_x; i++){
        for (let j = 0; j < numTiles_y; j++){
            let tile = getTile(i, j, level_color);
            if (tile.timer){
                tile.timer--;
                if (tile.timer == 0){
                    tile.timesUp();
                }
            }
        }
    }
}

function screenshake(){
    if(shakeAmount){
        shakeAmount--;
    }
    let shakeAngle = Math.random()*Math.PI*2;
	shakeX = Math.round(Math.cos(shakeAngle)*shakeAmount);
	shakeY = Math.round(Math.sin(shakeAngle)*shakeAmount);
}

function getClosestMonster(){
    let dist = 100; // arbitrary large distance
    var monster
    for (let i = 0; i < monsters.length; i++){
        if (player.tile.dist(monsters[i].tile) < dist && monsters[i] != player){
            dist = player.tile.dist(monsters[i].tile);
            monster = monsters[i];
        }
    }
    return monster
}

function boltTravel(direction, effect, start_tile){
    let newTile = start_tile;
    while (true){
        let testTile = newTile.getNeighbor(direction[0], direction[1]);
        if(testTile.passable){
            newTile = testTile;
            if (newTile.monster){
                newTile.monster.hit();
            }
            newTile.setEffect(effect);
        } else {
            break;
        }
    }
}

function initSounds(){
    sounds = {
        hit_player: new Audio('sounds/player_hit.wav'),
        hit_slime: new Audio('sounds/slime_hit.wav'),
        cast_tornado: new Audio('sounds/cast_tornado.wav'),
        collect_gem: new Audio('sounds/collect_gem.wav'),
        eh_something: new Audio('sounds/eh something.wav'),
        explosion: new Audio('sounds/explosion.wav'),
        lightning: new Audio('sounds/lightning.wav'),
        new_level: new Audio('sounds/new_level.wav'),
    };
}

function playSound(soundName){
    sounds[soundName].currentTime = 0;
    sounds[soundName].volume = .75;
    sounds[soundName].play();
}
